[ecs] Split launch type from deployment mode

JLineaweaver opened the pull request:
### What does this PR do?
ECS historically has had two launch types. We deploy the Datadog agent differently depending on launch type

- ECS on EC2 runs a daemon
  - Single agent per host
  - Gets all the tasks
  - Gets host info
- Fargate
  - Has to run in side car container
  - 1x agent per task
  - Only gets info about the task it's a part of

With the launch of [ECS Managed Instances ](https://aws.amazon.com/blogs/aws/announcing-amazon-ecs-managed-instances-for-containerized-applications/) we need to support the new launch type. At the time of writing this managed instances does not support scheduling the agent as a daemon. This means the only option we have now is treating it like fargate and deploying as a sidecar. Eventually if daemon scheduling is launched we want to make sure that we have the ability to deploy the agent that way as well.  

The goal of this PR is to separate deployment method from launch type. This will allow us in a follow up PR to support managed instances without locking ourselves into a side car approach.

### Motivation

Support ECS Managed Instances in a way that long term will allow both a sidecar deployment or a daemon deployment

### Describe how you validated your changes

We validated our changes by deploying both to ecs on ec2 and to ecs fargate. Both setups appeared to work correctly. We then validated trying to deploy not on 'auto' mode verifying that the wrong config does not work as intended.

### Additional Notes
For testing other features we recommend just deploying and verifying that anything built upon containers or that is expected to work with ECS still works fine.

JLineaweaver committed:
[ecs] Split launch type from deployment mode

JLineaweaver committed:
Remove unneeded import

JLineaweaver committed:
Fix deployment test

JLineaweaver committed:
Refactor the ecsfargate code within ecs

We want to refactor the fargate code to separate agent deployment
strategy vs ecs launch type. To do so we removed ecsfargate folder and
separated everything into just the ecs folder. We now share as much code
as possible but make daemon vs sidecar clear

JLineaweaver committed:
Fix unused param linting issues

JLineaweaver committed:
Fix additional linter issues

JLineaweaver committed:
Move initialization to correct call + cleanup comments

JLineaweaver committed:
Cleanup fargate codeowners

JLineaweaver committed:
Merge branch 'main' into jlineaweaver/exp-3

JLineaweaver committed:
Add release note

JLineaweaver committed:
Fix release note

JLineaweaver committed:
Merge branch 'main' into jlineaweaver/exp-3

dustmop reviewed (approved):
LGTM for agent-configuration owned files

JLineaweaver committed:
Always initialize v2 endpoint as a fallback

If doing sidecar mode on ec2 (which we don't recommend) we want to make
sure we always initialize the v2 endpoint

JLineaweaver commented:
/merge
